server/etcdserver/v3_server.go

```diff
@@ -237,7 +237,11 @@ func (s *EtcdServer) Compact(ctx context.Context, r *pb.CompactionRequest) (*pb.
 		// the hash may revert to a hash prior to compaction completing
 		// if the compaction resumes. Force the finished compaction to
 		// commit so it won't resume following a crash.
+		//
+		// `applySnapshot` sets a new backend instance, so we need to acquire the bemu lock.
+		s.bemu.RLock()
 		s.be.ForceCommit()
```

Comment:
It's interesting whether we have good test-coverage of 'swapping' backends in the middle of compaction.

The scheduledCompaction theoretically might finish on the 'old backends' why the recovered remains not compacted...

Reply:
Let's discuss & fix the race condition between `compaction & defragmentation & snapshot` and `applySnapshot` cases in separate issues/PRs. I will drive it.